plugins/modules/ndo_l3out_bgp_peer.py

```diff
@@ -0,0 +1,911 @@
+    private_as_controls = module.params.get("private_as_controls")
+    state = module.params.get("state")
+
+    mso_template = MSOTemplate(mso, "l3out", template, template_id)
+    mso_template.validate_template("l3out")
+    mso_template.check_template_when_name_is_provided(import_route_map)
+    mso_template.check_template_when_name_is_provided(export_route_map)
+    mso_template.check_template_when_name_is_provided(peer_prefix)
+
+    l3out_object = mso_template.get_l3out_object(uuid=l3out_uuid, name=l3out, fail_module=True)
+    node_group_object = mso_template.get_l3out_node_group(node_group, l3out_object.details, fail_module=True)
+
+    bgp_peer_objects = get_bgp_peer_by_address(mso_template, node_group_object.details.get("bgpPeers", []), ipv4_addr=ipv4_addr, ipv6_addr=ipv6_addr)
+
+    reference_dict = {
+        "peerPrefix": {
+            "name": "peerPrefixName",
+            "reference": "peerPrefixRef",
+            "type": "bgpPeerPrefixPol",
+            "template": "peerPrefixTemplateName",
+            "templateId": "peerPrefixTemplateId",
+        },
+        "importRouteMap": {
+            "name": "importRouteMapName",
+            "reference": "importRouteMapRef",
+            "type": "routeMap",
+            "template": "importRouteMapTemplateName",
+            "templateId": "importRouteMapTemplateId",
+        },
+        "exportRouteMap": {
+            "name": "exportRouteMapName",
+            "reference": "exportRouteMapRef",
+            "type": "routeMap",
+            "template": "exportRouteMapTemplateName",
+            "templateId": "exportRouteMapTemplateId",
+        },
+    }
+
+    if bgp_peer_objects and (ipv4_addr or ipv6_addr):
+        mso_template.update_config_with_template_and_references(bgp_peer_objects.details, reference_dict)
+        mso.existing = copy.deepcopy(bgp_peer_objects.details)  # Query a specific object
+        mso.previous = copy.deepcopy(bgp_peer_objects.details)  # Query a specific object
+    elif bgp_peer_objects:
+        mso.existing = [
+            mso_template.update_config_with_template_and_references(bgp_peer, reference_dict) for bgp_peer in bgp_peer_objects
+        ]  # Query all objects
+
+    if state != "query":
+        bgp_peer_path = "/l3outTemplate/l3outs/{0}/nodeGroups/{1}/bgpPeers/{2}".format(
+            l3out_object.index, node_group_object.index, bgp_peer_objects.index if bgp_peer_objects else "-"
+        )
+
+    ops = []
+
+    if state == "present":
+        peer_prefix_is_empty = check_if_all_elements_are_none(peer_prefix.values()) if peer_prefix else True  # peer_prefix is None or empty dict => True
+        import_route_map_is_empty = check_if_all_elements_are_none(import_route_map.values()) if import_route_map else True
+        export_route_map_is_empty = check_if_all_elements_are_none(export_route_map.values()) if export_route_map else True
+
+        mso_template_cache = MSOTemplateCache()
```

Comment:
We can do away with this if cache is moved to MSOTemplate itself and use mso_template object instantiated above?

Reply:
Done